solr/core/src/java/org/apache/solr/util/numeric/IntFloatDynamicMap.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util.numeric;
+
+import java.util.Arrays;
+
+import com.carrotsearch.hppc.IntFloatHashMap;
+import com.carrotsearch.hppc.cursors.FloatCursor;
+import com.carrotsearch.hppc.procedures.IntFloatProcedure;
+import org.apache.lucene.util.ArrayUtil;
+
+import static org.apache.solr.util.numeric.DynamicMap.mapExpectedElements;
+import static org.apache.solr.util.numeric.DynamicMap.threshold;
+import static org.apache.solr.util.numeric.DynamicMap.useArrayBased;
+
+public class IntFloatDynamicMap {
+  private int maxSize;
+  private IntFloatHashMap hashMap;
+  private float[] keyValues;
+  private float emptyValue;
+  private int threshold;
+
+  public IntFloatDynamicMap(int expectedMaxSize, float emptyValue) {
+    this.threshold = threshold(expectedMaxSize);
+    this.maxSize = expectedMaxSize;
+    this.emptyValue = emptyValue;
+    if (useArrayBased(expectedMaxSize)) {
+      upgradeToArray();
+    } else {
+      this.hashMap = new IntFloatHashMap(mapExpectedElements(expectedMaxSize));
+    }
+  }
+
+  private void upgradeToArray() {
+    keyValues = new float[maxSize];
+    if (emptyValue != 0.0f) {
+      Arrays.fill(keyValues, emptyValue);
+    }
+    if (hashMap != null) {
+      hashMap.forEach((IntFloatProcedure) (key, value) -> keyValues[key] = value);
+      hashMap = null;
+    }
+  }
+
+  private void growBuffer(int minSize) {
+    assert keyValues != null;
+    int size = keyValues.length;
+    keyValues = ArrayUtil.grow(keyValues, minSize);
+    if (emptyValue != 0.0f) {
+      for (int i = size; i < keyValues.length; i++) {
+        keyValues[i] = emptyValue;
+      }
+    }
+  }
+
+
+  public void set(int key, float value) {
+    if (keyValues != null) {
+      if (key >= keyValues.length) {
```

Comment:
Actually I missed here something risky: what if a large value key is provided? We have no protection here yet. Nothing prevents the user to put a key 500M. In this case we should not grow the array too large just for one key.
Maybe we should have a strong limit on the map capacity in the constructor.

Reply:
On a related note, do we even want to bother resizing this array vs size it appropriately up-front?